telemetry/Tracks/class-tracks-event.php

```diff
@@ -0,0 +1,275 @@
+<?php
+/**
+ * Telemetry: Tracks Event class
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry;
+
+use stdClass;
+use WP_Error;
+
+/**
+ * Class that creates and validates Tracks events.
+ *
+ * @since 3.12.0
+ */
+class Tracks_Event {
+	/**
+	 * Event name prefix.
+	 */
+	protected const EVENT_NAME_PREFIX = 'viptelemetry_';
+
+	/**
+	 * Event name regex.
+	 */
+	protected const EVENT_NAME_REGEX = '/^(([a-z0-9]+)_){1}([a-z0-9_]+)$/';
+
+	/**
+	 * Property name regex.
+	 */
+	protected const PROPERTY_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Variable containing the event's data or an error if one was encountered
+	 * during the event's creation.
+	 *
+	 * @var stdClass|WP_Error
+	 */
+	protected $data = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any properties included in the event.
+	 */
+	public function __construct( string $event_name, array $event_properties ) {
+		$event_data        = static::process_properties( $event_name, $event_properties );
+		$validation_result = static::get_event_validation_result( $event_data );
+
+		$this->data = $validation_result ?? $event_data;
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return stdClass|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function is_recordable() {
+		if ( ! Telemetry_System::is_wpadmin_telemetry_allowed() ) {
+			return false;
+		}
+
+		// Don't record events during unit tests and CI runs.
+		if ( 'wptests_capabilities' === wp_get_current_user()->cap_key ) {
+			return false;
+		}
+
+		if ( is_wp_error( $this->data ) ) {
+			return $this->data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Processes the event's properties to get them ready for validation.
+	 *
+	 * @param string $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any event properties to be processed.
+	 * @return stdClass The resulting event object with processed properties.
+	 */
+	protected static function process_properties(
+		string $event_name,
+		array $event_properties
+	): stdClass {
+		$event = (object) static::sanitize_properties_array( $event_properties );
+		$event = static::set_user_properties( $event );
+
+		// Set event name.
+		$event->_en = preg_replace(
+			'/^(?:' . static::EVENT_NAME_PREFIX . ')?(.*)/',
+			static::EVENT_NAME_PREFIX . '\1',
+			$event_name
+		) ?? '';
```

Comment:
It's not totally clear what the `preg_replace` is doing here.

We should also set a different prefix for local environments (e.g. `vip_dev_`.